### PR TITLE
[RFC] tests/ui: Fix indeterminism in Screen:wait

### LINKED ISSUE #2130: Functional test screen_basic_spec failing
Operating System:  Fedora 21

This  test fails 1  time out of every 4 runs or so.

screenshot from 2015-03-11 08 12 34
screenshot from 2015-03-11 08 11 56

If I can help in any other way, please do not hesitate to ask.

### PR DESCRIPTION
**Pull request by bfredl**
Fixes #2130 for me.

If the redraw stream looked like this
BAD
BAD
GOOD1
BAD2
GOOD3
stop() could be called when "GOOD1" is seen but  "BAD2" could already be
scheduled for processing (but not "GOOD3"), so that the loop exited in "BAD2"
without seeing "GOOD3", leading to a spurious failure.

Also avoids a unneccesary wait when the correct state already is reached.

### CODE REVIEW

**Comment by tarruda** [comment]
It may be better to fix in the tests, as I [explained](https://github.com/neovim/neovim/issues/2130#issuecomment-84610426) its possible that the bouncing problem is actually reporting a wrong assertion. If the redraw stream looks like this:

```
BAD
BAD
GOOD1
BAD2
GOOD3
```

Then its possible that the first `GOOD1` state is unintentionally caused by a previous input, in which case a `screen:expect` must be added before sending the input that results in `BAD2`.

**Comment by bfredl** [comment]
Ok, I'll see if an extra `expect` helps

**Comment by tarruda** [comment]
Do you think you could modify this to simply detect "bouncing" and print a warning to the user?

**Comment by bfredl** [comment]
Should it warn even if the test actually was successful, i e even if "GOOD3" was reached in the same redrawing batch?

**Comment by tarruda** [comment]
> Should it warn even if the test actually was successful, i e even if "GOOD3" was reached in the same redrawing batch?

Hmmm, maybe it should always warn if another state(even good) is received after a good state. This may force tests to be more strict regarding assumptions and placement of semaphores.

**Comment by bfredl** [comment]
After adding the #2130 fix, no other screen test shows this warning (at least in 100 iterations), so we should probably add it. Needs a traceback as well....

**Comment by tarruda** [comment]
> After adding the #2130 fix, no other screen test shows this warning (at least in 100 iterations), so we should probably add it. Needs a traceback as well....

:+1:

**Comment by bfredl** [comment]
BTW can't we just add `luarocks install inspect` to the installed packages? It is very handy for debugging any lua test, and is quite small and pure lua.

**Comment by bfredl** [comment]
I applied your fix as well. Not 100% happy with the wording of the warning, suggestions welcome.

File: test/functional/ui/screen.lua [hunk 1]
```diff
@@ -217,20 +217,38 @@ end
...
+  local failure_after_success = false
   local function notification_cb(method, args)
     assert(method == 'redraw')
     self:_redraw(args)
     err = check()
     checked = true
     if not err then
+      success_seen = true
       stop()
+    elseif success_seen then
```
Comment (tarruda) [inline test/functional/ui/screen.lua]: Will this trigger if two successes are received? Maybe it should be:

```lua
if success_seen then
  changes_after_success = true
else if not err then
  success_seen = true
  stop()
end
```

and the check:

```lua
if changes_after_success then
  print(...
```

Comment (tarruda) [inline test/functional/ui/screen.lua]: And the message:

```
Warning: Screen changes have been received after the expected state was seen...
```

(Adapt it so even two successes generate a warning)

**Comment by bfredl** [comment]
:+1: Though this generates a few warnings. For instance for `set_title` (as the tittle value is not part of the test :) ) but some `cursor_goto(0,0)` that seem interesting...

**Comment by bfredl** [comment]
Do you know how to truncate tracebacks "downwards"? I can get to the the call of `expect` with level=2, but then there are 20 lines of busted stuff....

**Comment by tarruda** [comment]
> :+1: Though this generates a few warnings. For instance for set_title (as the tittle value is not part of the test :) ) but some cursor_goto(0,0) that seem interesting...

If it generates warnings for some existing tests, then its best to leave as you first suggested for now(only a failure after a success trigger the warning)

**Comment by bfredl** [comment]
> If it generates warnings for some existing tests, then its best to leave as you first suggested for now(only a failure after a success trigger the warning)
:+1:

**Comment by bfredl** [comment]
@tarruda Updated. changed back to only warning on failure-after-success and fixed nicer traceback.

**Comment by tarruda** [comment]
@bfredl LGTM

If you repush the quickbuild ci should pass. I think it errored because of a indeterminism in job_wait tests that was fixed yesterday.

**Comment by bfredl** [comment]
Thanks, repushing.

**Comment by tarruda** [comment]
:+1: merged, thanks

**Pull request closed without merging**